wpilibc/src/main/native/cpp/Tracer.cpp

```diff
@@ -4,14 +4,226 @@
+    } else {
+      FRC_ReportWarning(
+          "Cannot update Tracer thread name after traces have been started");
+    }
+  }
+
+  // The network table that all data is published to
+  std::shared_ptr<nt::NetworkTable> m_rootTable;
+  // The stack of trace frames, every startTrace will add to this stack
+  // and every endTrace will remove from this stack
+  std::vector<std::string_view> m_traceStack;
+  // A map of trace names to the time they took to execute
+  wpi::StringMap<units::millisecond_t> m_traceTimes;
+  // A map of trace names to the time they started
+  wpi::StringMap<units::millisecond_t> m_traceStartTimes;
+  // A collection of all publishers that have been created,
+  // this makes updating the times of publishers much easier and faster
+  std::vector<std::pair<std::string_view, nt::DoublePublisher>> m_publishers;
+  // If the cycle is poisened, it will warn the user
+  // and not publish any data
+  bool m_cyclePoisened = false;
+  // If the tracer is disabled, it will not publish any data
+  // or do any string manipulation
+  bool m_disabled = false;
+  // If the tracer should be disabled next cycle
+  // and every cycle after that until this flag is set to false.
+  // Disabling is done this way to prevent disabling/enabling
+  // in the middle of a cycle
+  bool m_disableNextCycle = false;
+  // stack size is used to keep track of stack size
+  // even when disabled, calling `EndCycle` is important when
+  // disabled or not to update the disabled state in a safe manner
+  uint32_t m_stackSize = 0;
+};
+
+thread_local TracerState threadLocalState = TracerState();
+
+void Tracer::StartTrace(std::string_view name) {
+  // Call `AppendTraceStack` even if disabled to keep `m_stackSize` in sync
+  std::string stack = threadLocalState.AppendTraceStack(name);
+  if (!threadLocalState.m_disabled) {
+    auto now =
+        units::millisecond_t{frc::RobotController::GetFPGATime() / 1000.0};
+    threadLocalState.m_traceStartTimes.insert_or_assign(stack, now);
+  }
+}
+
+void Tracer::EndTrace() {
+  // Call `PopTraceStack` even if disabled to keep `m_stackSize` in sync
+  std::string stack = threadLocalState.PopTraceStack();
+  if (!threadLocalState.m_disabled) {
+    if (stack.empty()) {
+      FRC_ReportWarning("Trace ended without starting");
+      return;
+    }
+    auto now =
+        units::millisecond_t{frc::RobotController::GetFPGATime() / 1000.0};
+    auto startTime = threadLocalState.m_traceStartTimes.find(stack);
+    threadLocalState.m_traceTimes.insert_or_assign(stack,
+                                                   now - startTime->second);
+  }
+  // If the stack is empty, the cycle is over.
+  // Cycles can be ended even if the tracer is disabled
+  if (threadLocalState.m_stackSize == 0) {
+    threadLocalState.EndCycle();
+  }
+}
+
+void Tracer::EnableSingleThreadedMode() {
+  if (anyTracesStarted) {
+    FRC_ReportWarning(
+        "Cannot enable single-threaded mode after traces have been started");
```

Comment:
This won't work... anyTracesStarted is set to true when the thread local state is constructed, so this function will always be dead

Reply:
Moved anyTracesStarting to the `StartTrace` function, used to other langs "construct on use" paradigm for thread locals

Reply:
You should add a test to make sure `EnableSingleThreadedMode` works correctly in the context of a user project. (so any Tracer functions called before `RobotInit` in a robot project should be called first).